bliss/models/galaxy_net.py

```diff
@@ -1,174 +1,262 @@
+        residual_autoencoder: The second, "residual" AE.
+        residual_encoder: The encoder from the second, "residual" AE.
+        residual_decoder: The decoder from the second, "residual" AE.
+        slen: Side length of input images.
+        latent_dim: Latent dimension of encoded representation.
+        min_sd: Minimum sd for log-likelihood.
+        residual_delay_n_steps: Number of training steps before starting to
+            train residual model.
+        mse_residual_model_loss: If true, use MSE rather than log-likelihood
+            to train residual model.
+    """
 
     def __init__(
         self,
-        slen=53,
-        latent_dim=8,
-        hidden=256,
-        n_bands=1,
+        slen: int = 53,
+        latent_dim: int = 64,
+        n_bands: int = 1,
+        residual_delay_n_steps: int = 500,
+        mse_residual_model_loss: bool = False,
         optimizer_params: dict = None,
+        min_sd=1e-3,
     ):
+        """Initializer.
+
+        Args:
+            slen: (optional) Image side length.
+            latent_dim: (optional) Latent size of each autoencoder.
+            n_bands: (optional) Number of bands in image.
+            residual_delay_n_steps: (optional)
+                Number of training steps before starting to train residual model.
+            mse_residual_model_loss: (optional)
+                Use MSE instead of log-likelihood to train residual AE?
+            optimizer_params: (optional)
+                Parameters used to construct training optimizer.
+            min_sd: (optional)
+                Minimum sd for log-likelihood.
+        """
         super().__init__()
         self.save_hyperparameters()
 
-        self.enc = CenteredGalaxyEncoder(
-            slen=slen, latent_dim=latent_dim, hidden=hidden, n_bands=n_bands
+        self.main_encoder = CenteredGalaxyEncoder(slen=slen, latent_dim=latent_dim, n_bands=n_bands)
+        self.main_decoder = CenteredGalaxyDecoder(slen=slen, latent_dim=latent_dim, n_bands=n_bands)
+        self.main_autoencoder = nn.Sequential(self.main_encoder, self.main_decoder)
+
+        self.residual_encoder = CenteredGalaxyEncoder(
+            slen=slen, latent_dim=latent_dim, n_bands=n_bands
         )
-        self.dec = CenteredGalaxyDecoder(
-            slen=slen, latent_dim=latent_dim, hidden=hidden, n_bands=n_bands
+        self.residual_decoder = CenteredGalaxyDecoder(
+            slen=slen, latent_dim=latent_dim, n_bands=n_bands
         )
+        self.residual_autoencoder = nn.Sequential(self.residual_encoder, self.residual_decoder)
 
-        self.register_buffer("zero", torch.zeros(1))
-        self.register_buffer("one", torch.ones(1))
+        self.mse_residual_model_loss = mse_residual_model_loss
+
+        self.residual_delay_n_steps = residual_delay_n_steps
+        assert slen == 53, "Currently slen is fixed at 53"
+        self.slen = slen
+        assert latent_dim == 64, "Currently latent_dim is fixed at 64"
```

Comment:
I agree this is ugly. Right now the code is not adaptable to arbitrary galaxy "slen" or latent dimension sizes. The encoder consists of five downsampling layers, which take the image down to a `(32,1,1)` size representation. Thus, this will only work for images within a certain slen. Moreover, since I saw reduced performance adding linear layers after, there's no way to tweak the latent dimension size.

I'll try to think of a less ugly way to do this. Likely it will be:
-  Make the number of downsample layers depend on "slen". Specifically, add downsample layers until the image is reduced to `(2**L,1,1)`, where `L` is the number of layers.
-  Set `latent_dim` equal to `2*L`.

Reply:
Made this a new issue in #343, because it's not causing problems in our pipeline right now, but it should be addressed. I'll likely have to refactor the code some.